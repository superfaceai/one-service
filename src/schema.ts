import {
  detectSuperJson,
  META_FILE,
  loadSuperJson as sdkLoadSuperJson,
  NodeFileSystem,
  normalizeSuperJsonDocument,
} from '@superfaceai/one-sdk';
import { anonymizeSuperJson } from '@superfaceai/one-sdk/dist/core/events/reporter/utils';
import { NormalizedSuperJsonDocument } from '@superfaceai/ast';
import createDebug from 'debug';
import {
  GraphQLFieldConfig,
  GraphQLFieldConfigMap,
  GraphQLList,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
  printSchema,
} from 'graphql';
import { join as joinPath } from 'path';
import { DEBUG_PREFIX } from './constants';
import { load as loadProfile } from './profile';
import { generateProfileConfig, generateProfileTypes } from './schema.types';
import { hasFieldsDefined, sanitizedProfileName } from './schema.utils';

const debug = createDebug(`${DEBUG_PREFIX}:schema`);

export async function createSchema(): Promise<GraphQLSchema> {
  const superJson = await loadSuperJson();
  return await generate(superJson);
}

export async function loadSuperJson(): Promise<NormalizedSuperJsonDocument> {
  const superPath = await detectSuperJson(process.cwd(), NodeFileSystem);
  if (!superPath) {
    throw new Error('Unable to generate, super.json not found');
  }

  const loadedResult = await sdkLoadSuperJson(
    joinPath(superPath, META_FILE),
    NodeFileSystem,
  );

  const superJsonDocument = loadedResult.match(
    (v) => v,
    (err) => {
      throw new Error(`Unable to load super.json: ${err.formatShort()}`);
    },
  );

  return normalizeSuperJsonDocument(superJsonDocument);
}

export async function generate(
  superJson: NormalizedSuperJsonDocument,
): Promise<GraphQLSchema> {
  const queryFields: GraphQLFieldConfigMap<any, any> = {};
  const mutationFields: GraphQLFieldConfigMap<any, any> = {};

  for (const [profile, profileSettings] of Object.entries(superJson.profiles)) {
    debug(`generate start for ${profile}`);

    const loadedProfile = await loadProfile(superJson, profile);

    const profilePrefix = sanitizedProfileName(loadedProfile.ast);

    const { QueryType, MutationType } = await generateProfileTypes(
      profilePrefix,
      loadedProfile.ast,
      profileSettings,
      superJson.providers,
    );

    if (QueryType) {
      if (queryFields[profilePrefix]) {
        throw new Error(`Profile name collision in Query: ${profilePrefix}`);
      }

      queryFields[profilePrefix] = generateProfileConfig(
        QueryType,
        loadedProfile.ast,
      );
    }

    if (MutationType) {
      if (mutationFields[profilePrefix]) {
        throw new Error(`Profile name collision in Mutation: ${profilePrefix}`);
      }

      mutationFields[profilePrefix] = generateProfileConfig(
        MutationType,
        loadedProfile.ast,
      );
    }
  }

  queryFields['_superJson'] = superJsonFieldConfig(superJson);

  const schema = new GraphQLSchema({
    description: 'Superface.ai ❤️',
    query: new GraphQLObjectType({
      name: 'Query',
      description: "Profile's safe use-cases",
      fields: queryFields,
    }),
    mutation: hasFieldsDefined(mutationFields)
      ? new GraphQLObjectType({
          name: 'Mutation',
          description: "Profile's unsafe and idempotent use-cases",
          fields: mutationFields,
        })
      : undefined,
  });

  debug('generated schema:\n', printSchema(schema));

  return schema;
}

export function superJsonFieldConfig(
  superJsonDocument?: NormalizedSuperJsonDocument,
): GraphQLFieldConfig<any, any> {
  return {
    type: new GraphQLObjectType({
      name: 'SuperJson',
      fields: {
        profiles: {
          type: new GraphQLList(
            new GraphQLObjectType({
              name: 'ProfileInfo',
              fields: {
                name: {
                  type: GraphQLString,
                },
                version: {
                  type: GraphQLString,
                },
                providers: {
                  type: new GraphQLList(GraphQLString),
                },
              },
            }),
          ),
        },
        providers: {
          type: new GraphQLList(GraphQLString),
        },
      },
    }),
    resolve: async () => {
      const superJson = superJsonDocument ?? (await loadSuperJson());
      const anonymizedSuperJson = anonymizeSuperJson(superJson);

      return {
        profiles: Object.entries(anonymizedSuperJson.profiles).map(
          ([name, info]) => ({
            name,
            version: info.version,
            providers: info.providers.map((p) => p.provider),
          }),
        ),
        providers: anonymizedSuperJson.providers,
      };
    },
  };
}
