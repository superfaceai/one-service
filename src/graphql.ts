import { graphqlHTTP, OptionsData } from 'express-graphql';
import { IncomingMessage, ServerResponse } from 'http';
import { createSchema } from './schema';
import { GraphQLSchema } from 'graphql';
import { assertMapDocumentNode } from '@superfaceai/ast';

declare type Request = IncomingMessage & {
  url: string;
};
declare type Response = ServerResponse & {
  json?: (data: unknown) => void;
};

export type Middleware = (
  request: Request,
  response: Response,
) => Promise<void>;

export interface CreateGraphQLServerOptions
  extends Omit<OptionsData, 'schema'> {
  schema?: GraphQLSchema;
}

export async function createGraphQLMiddleware(
  options: CreateGraphQLServerOptions = {},
): Promise<Middleware> {
  options.schema = options.schema ?? (await createSchema());

  if (!isOptionsData(options)) {
    throw new Error('Property "schema" is missing');
  }

  return graphqlHTTP(options);
}

export function isOptionsData(
  options: CreateGraphQLServerOptions,
): options is OptionsData {
  return options.schema instanceof GraphQLSchema;
}
