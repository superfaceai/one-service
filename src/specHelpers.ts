import { SuperJson } from '@superfaceai/one-sdk';
import { join as joinPath } from 'path';

export function createSuperJson(profileFixtureName: string): SuperJson {
  return new SuperJson({
    profiles: {
      test: {
        file: joinPath(__dirname, 'fixtures', profileFixtureName),
        version: '1.0.0',
        providers: {
          mock: {},
        },
      },
    },
    providers: {
      mock: {
        security: [],
      },
    },
  });
}
