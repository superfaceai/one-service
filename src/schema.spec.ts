import { GraphQLSchema } from 'graphql';
import { generate } from './schema';
import { createSuperJson } from './specHelpers';

describe('schema', () => {
  describe('createSchema', () => {});

  describe('generate', () => {
    it('creates schema', async () => {
      const schema = generate(createSuperJson('test.supr'));

      await expect(schema).resolves.toBeInstanceOf(GraphQLSchema);
    });
  });

  describe('loadProfile', () => {});

  describe('capitalize', () => {});

  describe('camelize', () => {});
});
