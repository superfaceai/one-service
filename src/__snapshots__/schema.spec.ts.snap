// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema generate generates valid schema for profile without scope 1`] = `[]`;

exports[`schema generate generates valid schema for profile without scope 2`] = `
""""Superface.ai ❤️"""
schema {
  query: Query
}

"""Profile's safe use-cases"""
type Query {
  Profile: ProfileQuery
  _superJson: SuperJson
}

type ProfileQuery {
  UseCase(provider: ProfileProviderOption): ProfileUseCaseResult
}

"""
Wrapping type to handle many possible types returned as result by OneSDK
"""
type ProfileUseCaseResult {
  result: String
}

"""Provider configuration for OneSDK perform"""
input ProfileProviderOption {
  """Provider test configuration"""
  test: ProfileProviderOptiontestConfig
}

input ProfileProviderOptiontestConfig {
  parameters: ProfileProviderOptiontestProviderParameters
  security: ProfileProviderOptiontestProviderSecurity
}

"""Provider-specific parameters"""
input ProfileProviderOptiontestProviderParameters {
  """Parameter accepted by test"""
  param_no_default: String

  """Parameter accepted by test"""
  param_default: String
}

"""Provider-specific security"""
input ProfileProviderOptiontestProviderSecurity {
  """Security accepted by test"""
  api_key: TestApiKeySecurityValues

  """Security accepted by test"""
  basic: TestBasicSecurityValues

  """Security accepted by test"""
  bearer_token: TestBearerTokenSecurityValues

  """Security accepted by test"""
  digest: TestDigestSecurityValues
}

input TestApiKeySecurityValues {
  apikey: String
}

input TestBasicSecurityValues {
  username: String
  password: String
}

input TestBearerTokenSecurityValues {
  token: String
}

input TestDigestSecurityValues {
  username: String
  password: String
}

type SuperJson {
  profiles: [ProfileInfo]
  providers: [String]
}

type ProfileInfo {
  name: String
  version: String
  providers: [String]
}"
`;

exports[`schema generate generates valid schema for usecases mapped to mutation only 1`] = `[]`;

exports[`schema generate generates valid schema for usecases mapped to mutation only 2`] = `
""""Superface.ai ❤️"""
schema {
  query: Query
  mutation: Mutation
}

"""Profile's safe use-cases"""
type Query {
  _superJson: SuperJson
}

type SuperJson {
  profiles: [ProfileInfo]
  providers: [String]
}

type ProfileInfo {
  name: String
  version: String
  providers: [String]
}

"""Profile's unsafe and idempotent use-cases"""
type Mutation {
  ScopeName: ScopeNameMutation
}

type ScopeNameMutation {
  UnsafeUsecase(provider: ScopeNameProviderOption): ScopeNameUnsafeUsecaseResult
}

"""
Wrapping type to handle many possible types returned as result by OneSDK
"""
type ScopeNameUnsafeUsecaseResult {
  result: Int
}

"""Provider configuration for OneSDK perform"""
input ScopeNameProviderOption {
  """Provider test configuration"""
  test: ScopeNameProviderOptiontestConfig
}

input ScopeNameProviderOptiontestConfig {
  parameters: ScopeNameProviderOptiontestProviderParameters
  security: ScopeNameProviderOptiontestProviderSecurity
}

"""Provider-specific parameters"""
input ScopeNameProviderOptiontestProviderParameters {
  """Parameter accepted by test"""
  param_no_default: String

  """Parameter accepted by test"""
  param_default: String
}

"""Provider-specific security"""
input ScopeNameProviderOptiontestProviderSecurity {
  """Security accepted by test"""
  api_key: TestApiKeySecurityValues

  """Security accepted by test"""
  basic: TestBasicSecurityValues

  """Security accepted by test"""
  bearer_token: TestBearerTokenSecurityValues

  """Security accepted by test"""
  digest: TestDigestSecurityValues
}

input TestApiKeySecurityValues {
  apikey: String
}

input TestBasicSecurityValues {
  username: String
  password: String
}

input TestBearerTokenSecurityValues {
  token: String
}

input TestDigestSecurityValues {
  username: String
  password: String
}"
`;
